File: clients/java/src/main/java/org/aigen/oabp/model/Reward.java
package org.aigen.oabp.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.math.BigDecimal;

/**
 * The bounty attached to a mission: an {@code amount} in a given {@link Currency}.
 *
 * <p>The amount is modeled as {@link BigDecimal} (not {@code double}) so that token
 * and stablecoin quantities round-trip exactly. {@code AIGEN} is the protocol's
 * uncapped, off-chain points token; {@code USDC} is real value.
 *
 * @param amount   the gross reward quantity (before the protocol fee), never {@code null}
 * @param currency the settlement unit
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record Reward(
        @JsonProperty("amount") BigDecimal amount,
        @JsonProperty("currency") Currency currency) {

    /** The protocol fee, expressed as a fraction (0.5%). */
    public static final BigDecimal PROTOCOL_FEE_RATE = new BigDecimal("0.005");

    public Reward {
        if (amount == null) {
            amount = BigDecimal.ZERO;
        }
        if (currency == null) {
            currency = Currency.UNKNOWN;
        }
    }

    /**
     * The amount the winner actually receives after the 0.5% protocol fee is deducted.
     *
     * @return {@code amount * (1 - 0.005)}, with the same currency
     */
    public BigDecimal netAmount() {
        return amount.multiply(BigDecimal.ONE.subtract(PROTOCOL_FEE_RATE));
    }

    /**
     * The 0.5% protocol fee taken from this reward.
     *
     * @return {@code amount * 0.005}
     */
    public BigDecimal protocolFee() {
        return amount.multiply(PROTOCOL_FEE_RATE);
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/model/Reward.kt
package org.aigen.oabp.model

import kotlinx.serialization.Serializable
import java.math.BigDecimal

/**
 * The bounty attached to a mission: an [amount] in a given [Currency].
 *
 * The amount is a [BigDecimal] (not `Double`) so token and stablecoin quantities round-trip
 * exactly. `AIGEN` is the protocol's uncapped, off-chain points token; `USDC` is real value.
 *
 * @property amount the gross reward quantity (before the 0.5% protocol fee).
 * @property currency the settlement unit.
 */
@Serializable
public data class Reward(
    @Serializable(with = BigDecimalSerializer::class)
    val amount: BigDecimal,
    val currency: Currency,
) {
    /** The amount the winner actually receives after the 0.5% protocol fee: `amount * (1 - 0.005)`. */
    public val netAmount: BigDecimal
        get() = amount.multiply(BigDecimal.ONE.subtract(PROTOCOL_FEE_RATE))

    /** The 0.5% protocol fee taken from this reward: `amount * 0.005`. */
    public val protocolFee: BigDecimal
        get() = amount.multiply(PROTOCOL_FEE_RATE)

    public companion object {
        /** The protocol fee, expressed as a fraction (0.5%). */
        public val PROTOCOL_FEE_RATE: BigDecimal = BigDecimal("0.005")
    }
}
